DataManagementSystem/Client/ConsistencyChecks.py

```diff
@@ -0,0 +1,836 @@
+        if res['OK']:
+          present.update( res['Value']['Successful'] )
+          self.cachedReplicas.update( res['Value']['Successful'] )
+          notPresent.update( res['Value']['Failed'] )
+          break
+        else:
+          gLogger.error( "\nError getting replicas from FC, retry", res['Message'] )
+          time.sleep( 0.1 )
+    self.__write( ' (%.1f seconds)\n' % ( time.time() - startTime ) )
+
+    gLogger.info( "Found %d files with replicas and %d without" % ( len( present ), len( notPresent ) ) )
+    return list( present ), list( notPresent )
+
+  ################################################################################
+
+  def getReplicasPresenceFromDirectoryScan( self, lfns ):
+    """ Get replicas scanning the directories. Might be faster.
+    """
+
+    dirs = {}
+    present = []
+    notPresent = []
+    compare = True
+
+    for lfn in lfns:
+      dirN = os.path.dirname( lfn )
+      if lfn == dirN + '/':
+        compare = False
+      dirs.setdefault( dirN, [] ).append( lfn )
+
+    if compare:
+      self.__write( "Checking File Catalog for %d files from %d directories " % ( len( lfns ), len( dirs ) ) )
+    else:
+      self.__write( "Getting files from %d directories " % len( dirs ) )
+    startTime = time.time()
+
+    for dirN in sorted( dirs ):
+      startTime1 = time.time()
+      self.__write( '.' )
+      lfnsFound = self._getFilesFromDirectoryScan( dirN )
+      gLogger.verbose( "Obtained %d files in %.1f seconds" % ( len( lfnsFound ), time.time() - startTime1 ) )
+      if compare:
+        pr, notPr = self.__compareLFNLists( dirs[dirN], lfnsFound )
+        notPresent += notPr
+        present += pr
+      else:
+        present += lfnsFound
+
+    self.__write( ' (%.1f seconds)\n' % ( time.time() - startTime ) )
+    gLogger.info( "Found %d files with replicas and %d without" % ( len( present ), len( notPresent ) ) )
+    return present, notPresent
+
+  ################################################################################
+
+  @staticmethod
+  def __compareLFNLists( lfns, lfnsFound ):
+    """ return files in both lists and files in lfns and not in lfnsFound
+    """
+    present = []
+    notPresent = lfns
+    startTime = time.time()
+    gLogger.verbose( "Comparing list of %d LFNs with second list of %d" % ( len( lfns ), len( lfnsFound ) ) )
+    if lfnsFound:
+      setLfns = set( lfns )
+      setLfnsFound = set( lfnsFound )
+      present = list( setLfns & setLfnsFound )
+      notPresent = list( setLfns - setLfnsFound )
+
+    gLogger.verbose( "End of comparison: %.1f seconds" % ( time.time() - startTime ) )
+    return present, notPresent
+
+  def _getFilesFromDirectoryScan( self, dirs ):
+    """ calls dm.getFilesFromDirectory
+    """
+
+    level = gLogger.getLevel()
+    gLogger.setLevel( 'FATAL' )
+    res = self.dm.getFilesFromDirectory( dirs )
+    gLogger.setLevel( level )
+    if not res['OK']:
+      if 'No such file or directory' not in res['Message']:
```

Comment:
Since this is for v6r15, I would strongly encourage using the new error system, namely cmpError here :-)
https://github.com/DIRACGrid/DIRAC/blob/rel-v6r15/Core/Utilities/DErrno.py 

I will document all this in diracgrid